### PR TITLE
OCI: Assert that image hashes are equal across registries

### PR DESCRIPTION
**Pull request by sergiodj**
This commit implements a new test whose purpose is to assert that the
hashes for a certain image-arch-tag combination are equal across all
supported registries.

The idea is simple, but the implementation is a bit involved because
of the many corner cases I've had to deal with.

This test is making certain assumptions.  For example, it is
assuming that all supported architectures have been properly published
in all registries; if it notices that this is not the case, then it
errors out.  It is also only testing images from the "ubuntu"
namespace, because that's the only one that is published on both
registries.

Unfortunately, like the test to check if all supported architectures
are available, this test is also currently failing due to the problem
with our s390x 21.04 images.

### CODE REVIEW

File: oci-unit-tests/standalone-checks_test.sh
Comment (athos-ribeiro) [inline oci-unit-tests/standalone-checks_test.sh:140]: Should we always trust this assumption?
In the hypothetical scenario where we had just published a new image in a new namespace (or just a new tag), and the push operation failed for aws, this test could catch it in case we do not make this assumption. The  extra requests to dockerhub mighty be worth it to cover such corner cases.

I do understand this is not the purpose of this specific test though.

File: oci-unit-tests/standalone-checks_test.sh
Comment (athos-ribeiro) [inline oci-unit-tests/standalone-checks_test.sh:185]: Before this check, it would be nice if the file being handled here is a manifest list and not something else such as a json error or a image manifest for some specific arch.

**Review by athos-ribeiro — COMMENTED** [review]
Looks good!

Added a few questions and suggestions inline.

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:24]: You're right, this message is duplicated; the same message already exists in the `test_all_supported_architectures_are_available`, so I'm deleting this one.

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:34]: Good point.  I can check whether `mktemp` succeeded in the `oneTimeSetUp`.

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:119]: Unfortunately they need to be set because the scripts in the `utils` directory don't accept anonymous queries (which is kind of a good thing, because non-anonymous queries have a higher API ratelimit).

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:140]: I thought about this while writing this code.  I was also not happy to default to a specific registry for this query, but I thought that if we want to make sure that the same images are being published in all supported registries, then the best place to do this would be in another test (which I was planning to write).  But as I said I agree with your point, so I will see about making this part more robust.

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:185]: Right, but my understanding is that if we call the `jq` command on the supposed valid manifest list and it errors out, this means that we're dealing with a broken manifest list.  Or do you have something else in mind here?

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh]: OK, I've pushed something that I think is better.  Please have a look :-).

**Comment by sergiodj** [comment]
Thanks for the review, @athos-ribeiro.  I think I've addressed most of your comments, so let me know what you think.

BTW, I didn't post the current output of the test, so here is:

```
E: Could not obtain digest from manifest list for aws image ubuntu/apache2:2.4-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/apache2:2.4-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/apache2:2.4-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/apache2:2.4-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/apache2:edge (s390x)                                                                                                      
E: Could not obtain digest from manifest list for docker image ubuntu/apache2:edge (s390x)                                                                                                   
E: Could not obtain digest from manifest list for aws image ubuntu/apache2:latest (s390x)                                                                                                    
E: Could not obtain digest from manifest list for docker image ubuntu/apache2:latest (s390x)                                                                                                 
E: Could not obtain digest from manifest list for aws image ubuntu/cortex:1.7-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/cortex:1.7-21.04_beta (s390x)                                                                                          
E: Could not obtain digest from manifest list for aws image ubuntu/cortex:1.7-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/cortex:1.7-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/cortex:edge (s390x)                                                                                                       
E: Could not obtain digest from manifest list for docker image ubuntu/cortex:edge (s390x)                                                                                                    
E: Could not obtain digest from manifest list for aws image ubuntu/cortex:latest (s390x)                                                                                                     
E: Could not obtain digest from manifest list for docker image ubuntu/cortex:latest (s390x)                                                                                                  
E: Could not obtain digest from manifest list for aws image ubuntu/grafana:7.4-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/grafana:7.4-21.04_beta (s390x)                                                                                         
E: Could not obtain digest from manifest list for aws image ubuntu/grafana:7.4-21.04_edge (s390x)                                                                                            
E: Could not obtain digest from manifest list for docker image ubuntu/grafana:7.4-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/grafana:edge (s390x)                                                                                                      
E: Could not obtain digest from manifest list for docker image ubuntu/grafana:edge (s390x)                                                                                                   
E: Could not obtain digest from manifest list for aws image ubuntu/grafana:latest (s390x)                                                                                                    
E: Could not obtain digest from manifest list for docker image ubuntu/grafana:latest (s390x)                                                                                                 
E: Could not obtain digest from manifest list for aws image ubuntu/memcached:1.6-21.04_beta (s390x)    
E: Could not obtain digest from manifest list for docker image ubuntu/memcached:1.6-21.04_beta (s390x)    
E: Could not obtain digest from manifest list for aws image ubuntu/memcached:1.6-21.04_edge (s390x)      
E: Could not obtain digest from manifest list for docker image ubuntu/memcached:1.6-21.04_edge (s390x)      
E: Could not obtain digest from manifest list for aws image ubuntu/memcached:edge (s390x)                                                                                                    
E: Could not obtain digest from manifest list for docker image ubuntu/memcached:edge (s390x)                                                                                                 
E: Could not obtain digest from manifest list for aws image ubuntu/memcached:latest (s390x)                                                                                                  
E: Could not obtain digest from manifest list for docker image ubuntu/memcached:latest (s390x)    
E: Could not obtain digest from manifest list for aws image ubuntu/mysql:8.0-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/mysql:8.0-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/mysql:8.0-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/mysql:8.0-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/mysql:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/mysql:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/mysql:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/mysql:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/nginx:1.18-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/nginx:1.18-21.04_beta (s390x)                                                                                 [35/1901]
E: Could not obtain digest from manifest list for aws image ubuntu/nginx:1.18-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/nginx:1.18-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/nginx:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/nginx:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/nginx:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/nginx:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/postgres:13.1-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/postgres:13.1-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/postgres:13.1-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/postgres:13.1-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/postgres:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/postgres:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/postgres:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/postgres:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus:2.25-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus:2.25-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus:2.25-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus:2.25-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus-alertmanager:0.21-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus-alertmanager:0.21-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus-alertmanager:0.21-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus-alertmanager:0.21-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus-alertmanager:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus-alertmanager:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/prometheus-alertmanager:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/prometheus-alertmanager:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/redis:6.0-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/redis:6.0-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/redis:6.0-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/redis:6.0-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/redis:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/redis:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/redis:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/redis:latest (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/telegraf:1.17-21.04_beta (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/telegraf:1.17-21.04_beta (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/telegraf:1.17-21.04_edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/telegraf:1.17-21.04_edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/telegraf:edge (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/telegraf:edge (s390x)
E: Could not obtain digest from manifest list for aws image ubuntu/telegraf:latest (s390x)
E: Could not obtain digest from manifest list for docker image ubuntu/telegraf:latest (s390x)
ASSERT:Not all hashes are equal

Ran 1 test.

FAILED (failures=1)
```

As you can see, it's correctly detecting that the 21.04 s390x images are not present anywhere (that's why it's failed).  Other than that, all hashes were verified to be equal.

File: oci-unit-tests/standalone-checks_test.sh
Comment (sergiodj) [reply oci-unit-tests/standalone-checks_test.sh:145]: Yeah, I thought about doing that but then got discouraged because it will require some massaging of the data (one of the downsides of using shell script for these tests).  I'll see what I can do.

**Comment by sergiodj** [comment]
@athos-ribeiro Thanks for the review.  I have force-pushed some changes that should address your requests; I'd appreciate if you could take a second (third?  fourth?!) look, please :-).

Also, since @paride mentioned that he's reviewing this PR, I'm going to wait for his ACK before I merge it.

**Comment by sergiodj** [comment]
A kind ping to @paride just so that this PR doesn't fall through the cracks :-).

**Review by paride — APPROVED** [review]
Mmh I thought I already commented on this, but apparently I forgot to actually submit the review.

Overall this is very nicely done and works fine, I just have one comment: the standalone tests succeed if a nonexisting package is specified via `DOCKER_PACKAGE`. I think the test should fail instead, so should one of our images disappear we'd get a failure here.

I'm approving this already as I think it's ready to me merged, leaving the actual merge to you in case you want to implement the change above as part of this PR.

**Comment by sergiodj** [comment]
Thanks for the review, @paride.

The problem with detecting whether `DOCKER_PACKAGE` exists or not is that we have images that are present in one namespace but not the other (for example, `telegraf` exists on `ubuntu` but not on `lts`).

I implemented a way to perform this detection properly and only error out if `DOCKER_PACKAGE` has not been found anywhere.  It's a bit convoluted (but then again, this whole script is), but works.

I will go ahead and merge this PR, but if you and @athos-ribeiro have the time it'd be great to double check the new code.  Also, I have two questions for you:

1) Is it a problem if the script prints something on `stderr`?  This new check that I've implemented prints a warning if the image hasn't been found in a particular `registry/namespace` combination.  This means that, for images that are present in one namespace but not the other, a warning will always be printed.

2) Do you think it'd be possible to add these new tests to our CI?  :-)

Thanks!

**Pull request merged**
